docs/Customization/Custom-Core.rst

```diff
@@ -0,0 +1,267 @@
+Make sure to read :ref:`node_types` to check out what type of nodes Chipyard supports and their parameters!
+
+Also, by default, there are boundary buffers for both master and slave connections to the bus when they are leaving the tile, and you
+can override the following two functions to control how to buffer the bus requests/responses:
+
+.. code-block:: scala
+
+    protected def makeMasterBoundaryBuffers(implicit p: Parameters): TLBuffer
+    protected def makeSlaveBoundaryBuffers(implicit p: Parameters): TLBuffer
+
+You can find more information on ``TLBuffer`` in :ref:`diplomatic_widgets`.
+
+Create Implementation Class
+---------------------------
+
+The implementation class contains the parameterized, actual hardware that depends on the values resolved by the Diplomacy
+framework according to the info provided in the Tile class. This class will normally contains Chisel RTL codes, and if your
+core is in Verilog, you will need to put the black box class you created in the first step here and connect it with the buses
+and other components. No Diplomacy/TileLink code should be in this class; you should only connect the IO signals in TileLink
+interfaces or other diplomatically defined components, which are located in the tile class. 
+
+The implementation class for your core is of the following form:
+
+.. literalinclude:: ../../generators/chipyard/src/main/scala/example/TutorialTile.scala
+    :language: scala
+    :start-after: DOC include start: Implementation class
+    :end-before: DOC include end: Implementation class
+
+If you create an AXI4 node (or equivalents), you will need to connect them to your core. You can connect a port like this:
+
+.. literalinclude:: ../../generators/chipyard/src/main/scala/example/TutorialTile.scala
+    :language: scala
+    :start-after: DOC include start: AXI4 connect
+    :end-before: DOC include end: AXI4 connect
+
+Connect Interrupt
+-----------------
+
+Chipyard allows a tile to either receive interrupts from other devices or initiate interrupts to notify other cores/devices. 
+In the tile that inherited ``SinksExternalInterrupts``, one can create a ``TileInterrupts`` object (a Chisel bundle) and 
+call ``decodeCoreInterrupts`` with the object as the argument. Note that you should call this function in the implementation
+class since it returns a Chisel bundle used by RTL code. You can then read the interrupt bits from the object.
+The definition of ``TileInterrupts`` is 
+
+.. code-block:: scala
+
+    class TileInterrupts(implicit p: Parameters) extends CoreBundle()(p) {
+      val debug = Bool() // debug interrupt
+      val mtip = Bool() // Machine level timer interrupt
+      val msip = Bool() // Machine level software interrupt
+      val meip = Bool() // Machine level external interrupt 
+      val seip = usingSupervisor.option(Bool()) // Valid only if supervisor mode is supported
+      val lip = Vec(coreParams.nLocalInterrupts, Bool())  // Local interrupts
+    }
```

Comment:
Instead of having Rocketchip code that is likely to change, can we just put a reference to the relevant file in the Rocketchip repo?

Reply:
Unfortunately, our ReadTheDocs system does not yet initialize submodules in the repo (its supported, but someone needs to figure out how to make it work). A HTML link is the best you can do for now.

Reply:
HTML would be fine with me

Reply:
I would add some HTML links along with the code with my comments.

Reply:
Are you going to replace this inlined code with links?

Reply:
No. I am planning to add it beside the code since the original code does not have the comments. See my comment below.